Fix `update_all` producing a broken query when using more than one join on the target:

Edouard-chin opened the pull request:
- Fix #54560
- `update_all` would produce a broken query when using more than one JOIN on the target.

```ruby
Developer.join(:computer, :mentor).update_all(ready: true)
```

### Before

```sql
UPDATE developers SET ready = true FROM computers JOIN mentors on mentors.id = developers.mentor_id WHERE developer.computer_id = computer.id; # We can't make a JOIN on the developer table, this should in the FROM clause.
# We can't make a join on the target, that's what the FROM clause is used for.
```

### After

```sql
UPDATE developers SET ready = true FROM computers, mentors WHERE developer.computer_id = computer.id AND developers.mentor_id = mentors.id;
```

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

Edouard-chin committed:
Fix `update_all` producing a broken query:

- Fix #54560
- `update_all` would produce a broken query when using more than one
  JOIN on the target.

  ```ruby
  Developer.join(:computer, :mentor).update_all(ready: true)
  ```

  Before

  ```sql
  UPDATE developers SET ready = true FROM computers JOIN mentors on mentors.id = developers.mentor_id WHERE developer.computer_id = computer.id;
  # We can't make a JOIN on the developer table, this should in the FROM clause.
  ```

  After

  ```sql
  UPDATE developers SET ready = true FROM computers, mentors WHERE developer.computer_id = computer.id AND developers.mentor_id = mentors.id;
  ```

ghiculescu commented:
This fixes the crashes in our test suite that led to https://github.com/rails/rails/issues/54560

But at least one of the tests that was crashing, is now failing. The relevant `update_all` is updating 0 rows. You can simulate this in the initial test script pointed to this commit:

```ruby
    assert_equal 1, post.likes.update_all(seen: true) # fine
    assert_equal 1, post.likes.left_joins(:reply).update_all(seen: true) # 💥
    assert_equal 1, post.likes.for_post(post).update_all(seen: true) # fine
```

It looks like the left outer join has been replaced with an inner join?

<img width="711" alt="image" src="https://github.com/user-attachments/assets/8c3b6082-2e80-440b-b4db-2c4f2a18207b" />
<img width="722" alt="image" src="https://github.com/user-attachments/assets/36dfc792-f6fb-4785-9500-702c5c5abb8b" />

byroot commented:
Yeah, I'm tempted to just give up on this. But yeah, let's give it one more try. Let me know once you handle the left join.

Edouard-chin committed:
Modify `update_all` to use a sub select query in some cases:

- In https://github.com/rails/rails/commit/216b37bd8ca742b719e0ed1bd8d619560a3c5fd0,
  we attempted a fix let `update_all` use a subselect query instead
  of the `UPDATE FROM` scheme when a LEFT JOIN is used.

  This fix doesn't apply however for leading joings that get
  produced by a `through :table` which first creates
  a INNER JOIN, and then other joins given by the user.
  So such code `Post.through_comments.left_joins(:author)` will
  continue to try using the `UPDATE FROM` scheme and this won't work
  (see the commit mentioned above why).

  The fix in this commit checks whether there is a left join on the
  target, no matter whether it's the first join, and in such case
  makes `update_all` uses a subselect query.

Edouard-chin commented:
> let's give it one more try.

🤞 

I pushed a commit with the change related to the left_join issue with some explanation.

byroot commented on activerecord/lib/arel/visitors/sqlite.rb:
```diff
@@ -58,7 +64,7 @@ def prepare_update_statement(o)
             # for the first join:
             #   UPDATE table SET .. FROM joined_table WHERE ...
-            (o.relation.right.first.is_a?(Arel::Nodes::InnerJoin))
+            (o.relation.right.all? { |join| join.is_a?(Arel::Nodes::InnerJoin) || join.right.expr.right.relation != o.relation.left })
```
I'm not sure I understand that `||` condition, could you explain what this is trying to match for?

Edouard-chin replied:
When we have a join on the target (the table we are updating), the joined table has to be in the `FROM` clause for the query to be valid. This is what the first commit in this PR fixes.

By introducing this fix, we now hit the flexibility issue you previously mentioned when a left outer join is used. 
What this `||` checks, is whether there is any left joins on the target which we have to workaround and resort to making a subselect query instead.

This check is only required on left joins on the target (as we have to translate that join into a FROM and we hit the flexibility issue). Issuing a left outer join on another table is fine since we don't have to translate it for the query to be syntax valid.

byroot replied:
> What this `||` checks, is whether there is any left joins on the target which we have to workaround and resort to making a subselect query instead.

That is where I'm confused, this `if` condition is for case where we don't have to fallback to a a subquery. So why `||`, wasn't this meant to be `&&`? (Sorry, I'm not the best with Arel API)
